src/lib.rs

```diff
@@ -92,48 +95,58 @@ fn get_compact_symbol_table_impl(
     binary_data: &[u8],
     debug_data: &[u8],
     breakpad_id: &str,
-) -> Option<compact_symbol_table::CompactSymbolTable> {
+) -> Result<compact_symbol_table::CompactSymbolTable> {
     let mut reader = Cursor::new(binary_data);
-    if let Ok(hint) = goblin::peek(&mut reader) {
-        match hint {
-            Hint::Elf(_) => {
-                return elf::get_compact_symbol_table(binary_data, breakpad_id);
-            }
-            Hint::Mach(_) => {
-                return macho::get_compact_symbol_table(binary_data, breakpad_id);
-            }
-            Hint::MachFat(_) => {
-                let multi_arch = mach::MultiArch::new(binary_data).ok()?;
-                for fat_arch in multi_arch.iter_arches().filter_map(Result::ok) {
-                    let arch_slice = fat_arch.slice(binary_data);
-                    if let Some(table) = macho::get_compact_symbol_table(arch_slice, breakpad_id) {
-                        return Some(table);
-                    }
+    let mut first_error = None;
+    match goblin::peek(&mut reader)? {
+        Hint::Elf(_) => elf::get_compact_symbol_table(binary_data, breakpad_id),
+        Hint::Mach(_) => macho::get_compact_symbol_table(binary_data, breakpad_id),
+        Hint::MachFat(_) => {
+            let multi_arch = mach::MultiArch::new(binary_data)?;
+            for fat_arch in multi_arch.iter_arches().filter_map(std::result::Result::ok) {
+                let arch_slice = fat_arch.slice(binary_data);
+                let table = macho::get_compact_symbol_table(arch_slice, breakpad_id);
+                if table.is_err() {
```

Comment:
Another opportunity to use `match`:

```rust
match macho::get_compact_symbol_table(arch_slice, breakpad_id) {
    Ok(table) => return table,
    Err(e) => first_error = e,
}
```

Also, you're overwriting `first_error` even when it's already set, so a better name for the variable would be `last_error`. :)